### PR TITLE
states: add step state base class

### PR DESCRIPTION
**Pull request by cmatsuoka**
The context of a step is preserved to detect state changes between
executions and decide if new runs can be skipped. This change adds the
base class for step states.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_parts/state_manager/step_state.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any, Dict, Set
+
+from pydantic import Field
+from pydantic_yaml import YamlModel  # type: ignore
+
+
+class StepState(YamlModel, ABC):
+    """The context used to run a step for a part.
```
Comment (facundobatista) [inline craft_parts/state_manager/step_state.py:29]: "is used"?

File: craft_parts/state_manager/step_state.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    The step state contains environmental and project-specific configuration
+    data collected at step run time. Those properties are used to decide whether
+    the step should run again on a new lifecycle execution.
+    """
+
+    part_properties: Dict[str, Any] = Field({}, alias="properties")
```
Comment (facundobatista) [inline craft_parts/state_manager/step_state.py:36]: Why will we have two ways to refer to this? Wouldn't it make the learning more complex?

File: craft_parts/state_manager/step_state.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        :return: The newly created dictionary.
+        """
+        return self.dict(by_alias=True)
+
+    def write(self, filepath: Path) -> None:
+        """Write this state to disk."""
+        os.makedirs(filepath.parent, exist_ok=True)
+        yaml_data = self.yaml(by_alias=True)
+        with open(filepath, "w") as f:
+            f.write(yaml_data)
```
Comment (facundobatista) [inline craft_parts/state_manager/step_state.py:106]: easier to just do `filepath.write_text(yaml_data)`

File: craft_parts/state_manager/step_state.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+
+def _get_differing_keys(dict1, dict2) -> Set[str]:
```
Comment (facundobatista) [inline craft_parts/state_manager/step_state.py:109]: Please provide a docstring and extensive tests for this function.

Thanks!

File: craft_parts/state_manager/step_state.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    differing_keys = set()
+    for key, dict1_value in dict1.items():
+        dict2_value = dict2.get(key)
+        if dict1_value != dict2_value:
+            differing_keys.add(key)
```
Comment (facundobatista) [inline craft_parts/state_manager/step_state.py:114]: For the following case, should `foobar` be included in the `differing_keys` set?

- dict1: `{'foobar': None}`
- dict2: `{}`

**Review by facundobatista — CHANGES_REQUESTED** [review]
Nice! Added some comments inline, thanks!!!

File: craft_parts/state_manager/step_state.py (hunk 2 above)
Comment (cmatsuoka) [reply craft_parts/state_manager/step_state.py:36]: Changing to call it `part-properties` in the state file.

File: craft_parts/state_manager/step_state.py (hunk 5 above)
Comment (cmatsuoka) [reply craft_parts/state_manager/step_state.py:114]: It shouldn't, will document the rules.

File: craft_parts/state_manager/step_state.py (hunk 1 above)
Comment (cmatsuoka) [reply craft_parts/state_manager/step_state.py:29]: Rephrased that, we don't actually "use" a context to run a step.

File: craft_parts/state_manager/step_state.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    def marshal(self) -> Dict[str, Any]:
+        """Create a dictionary containing the part state data.
+
+        :return: The newly created dictionary.
+        """
+        return self.dict(by_alias=True)
+
+    def write(self, filepath: Path) -> None:
+        """Write this state to disk."""
+        os.makedirs(filepath.parent, exist_ok=True)
```
Comment (sergiusens) [inline craft_parts/state_manager/step_state.py:102]:
```suggestion
        filepath.parent.mkdir(parents=True, exist_ok=True)
```

File: craft_parts/state_manager/step_state.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The step state preserves step execution context information."""
+
+import os
```
Comment (sergiusens) [inline craft_parts/state_manager/step_state.py:19]:
```suggestion
```

**Review by sergiusens — APPROVED** [review]
Looks good, just one suggestion

**Review by sergiusens — APPROVED** [review]

File: craft_parts/state_manager/step_state.py (hunk 5 above)
Comment (cjp256) [reply craft_parts/state_manager/step_state.py:117]: _get_differing_keys can be simplified to:

diff_items = (dict1.items() - dict2.items()) | (dict2.items() - dict1.items())`
return [i[0] for i in diff_items]

File: craft_parts/state_manager/step_state.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        )
+
+    def marshal(self) -> Dict[str, Any]:
+        """Create a dictionary containing the part state data.
+
+        :return: The newly created dictionary.
+        """
+        return self.dict(by_alias=True)
+
+    def write(self, filepath: Path) -> None:
```
Comment (cjp256) [inline craft_parts/state_manager/step_state.py:99]: In craft-providers i'm using `load()` classmethod and `save()` instance method for the names of marshaling/unmarshaling to/from disk.  I don't care what we pick, but we should probably pick terms and be consistent like with marshal/unmarshal. :D

**Review by cjp256 — APPROVED** [review]

File: craft_parts/state_manager/step_state.py (hunk 8 above)
Comment (cmatsuoka) [reply craft_parts/state_manager/step_state.py:99]: It's a good point, I'll check that and refactor in a follow-up.

File: craft_parts/state_manager/step_state.py (hunk 5 above)
Comment (facundobatista) [reply craft_parts/state_manager/step_state.py:117]: Only if all the values from both dicts are hashable. Signature type hints say values can be `Any`.

File: tests/unit/state_manager/test_step_state.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            [{}, {"b": 2}, {"b"}],
+            [{"a": 1}, {"b": 2}, {"a", "b"}],
+            [{"a": None}, {}, set()],
+            [{}, {"b": None}, set()],
+            [{"a": None}, {"b": None}, set()],
+            [{"a": 1}, {"a": 1}, set()],
+            [{"a": None}, {"a": 1}, {"a"}],
+            [{"a": 1}, {"a": 1, "b": 2}, {"b"}],
+            [{"a": 1, "b": 2}, {"a": 1}, {"b"}],
+        ],
```
Comment (facundobatista) [inline tests/unit/state_manager/test_step_state.py:155]: Missing at least the case of same key with different not-None values.

**Review by facundobatista — APPROVED** [review]
Great work!! Approved, but please take a look to the comment in the test function.

Thanks!

File: tests/unit/state_manager/test_step_state.py (hunk 9 above)
Comment (cmatsuoka) [reply tests/unit/state_manager/test_step_state.py:155]: Adding this test.

**Pull request merged**